enclave/core/sgx/context.inc

```diff
@@ -147,5 +147,12 @@ oe_continue_execution:
     add     $(ABI_REDZONE_BYTE_SIZE + 8), %rsp
 
     // Jump to the target.
-    jmp     *-(ABI_REDZONE_BYTE_SIZE + 8)(%rsp)
+    // Since no register is free, we cannot load the target into a register and
+    // then issue an lfence. Instead, we use retq (return) instruction to carry
+    // out a jump. We first push the target address onto the stack, issue an
+    // lfence, and then do a retq. The retq will pop the target from the stack
+    // and then jump to it.
+    pushq     -(ABI_REDZONE_BYTE_SIZE + 8)(%rsp)
+    lfence
```

Comment:
`ret` instruction loads (reads) from memory (stack), then jump. A `lfence` instruction inserted before `ret` might not "cut" the potential LVI flow targeting the top of the stack.  

The `lfence` instruction does prevent speculative execution of `ret` before `pushq` instruction on line 155 retires (any exception on that instruction will be delivered before execution of next instruction), thus "cuts" the potential LVI flow targeting `-(ABI_REDZONE_BYTE_SIZE + 8)(%rsp)` address.  The `pushq` instruction on line 155 also stores (writes) to the top of the stack. But a potential attack might be able to manipulate the r/w permission of top of the stack location and only cause an exception when load from that location. In that case, the load part of the `ret` instruction will trigger exception and lead to potential LVI attack.

If the assembler-based LVI mitigation is enabled, according to Intel's [LVI Deep Dive Whitepaper](https://software.intel.com/security-software-guidance/deep-dives/deep-dive-load-value-injection#specialinstructions ), the assembler is supposed to replace `ret` instruction with:

SHL QWORD PTR [rsp], 0 
LFENCE 
RET 

The instruction sequence loads from target address, lfence, then load/jump. According tot he white paper, it "cuts" the potential LVI flow targeting the top of the stack. 

Relying on the assembler-based mitigation, the following instruction sequence should be sufficient.

    pushq     -(ABI_REDZONE_BYTE_SIZE + 8)(%rsp)
    ret

The white paper also mentions similar technique for indirect branch through memory operand:

XOR QWORD PTR <some GPR> , [rsi] 
XOR QWORD PTR <some GPR> , [rsi] 
LFENCE
JMP QWORD PTR [rsi] 

I wonder instead of using XOR to load the target memory, but causing flag change, whether `SHL QWORD PTR [rsi] , 0` also works.

Reply:
After converting the indirect `jmp` to `ret`, the assembler (when enabling LVI mitigation) should take care of what @bodzhang described. No need to manually insert `lfence`.

Reply:
Thanks for the explanation @bodzhang. I removed the extra lfence.
Note, the assembler transforms retq instruction as below:
```asm
    notq (%rsp)
    lfence
    retq
```